cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -0,0 +1,455 @@
+    auto const schema_idx = output_column_schemas[col_idx];
+    auto const& dtype     = output_dtypes[col_idx];
+    // Only comparable types except fixed point are supported.
+    if (cudf::is_compound(dtype) && dtype.id() != cudf::type_id::STRING) {
+      // placeholder only for unsupported types.
+      columns.push_back(cudf::make_numeric_column(
+        data_type{cudf::type_id::BOOL8}, total_rows, rmm::device_buffer{}, 0, stream, mr));
+      columns.push_back(cudf::make_numeric_column(
+        data_type{cudf::type_id::BOOL8}, total_rows, rmm::device_buffer{}, 0, stream, mr));
+      continue;
+    }
+    auto [min_col, max_col] =
+      cudf::type_dispatcher<dispatch_storage_type>(dtype, stats_col, schema_idx, dtype, stream, mr);
+    columns.push_back(std::move(min_col));
+    columns.push_back(std::move(max_col));
+  }
+  auto stats_table = cudf::table(std::move(columns));
+
+  // Converts AST to StatsAST with reference to min, max columns in above `stats_table`.
+  cudf::io::parquet::detail::stats_expression_converter const stats_expr{
+    filter.value().get(), static_cast<size_type>(output_dtypes.size())};
+
+  // Filter the input table using AST expression
+  return cudf::detail::compute_column(stats_table, stats_expr.get_stats_expr().get(), stream, mr);
+}
+
+std::vector<std::vector<cudf::io::text::byte_range_info>>
+aggregate_reader_metadata::get_filter_columns_data_pages(
+  cudf::column_view input_rows,
+  cudf::host_span<std::vector<size_type> const> row_group_indices,
+  host_span<data_type const> output_dtypes,
+  host_span<int const> output_column_schemas,
+  rmm::cuda_stream_view stream) const
+{
+  if (row_group_indices.empty()) { return {}; }
+
+  CUDF_EXPECTS(input_rows.type().id() == cudf::type_id::BOOL8,
+               "Input row bitmask should be of type BOOL8");
+
+  auto const host_bitmask = [&] {
+    auto const num_bitmasks = num_bitmask_words(input_rows.size());
+    if (input_rows.nullable()) {
+      return cudf::detail::make_host_vector_sync(
+        device_span<bitmask_type const>(input_rows.null_mask(), num_bitmasks), stream);
+    } else {
+      auto bitmask = cudf::detail::make_host_vector<bitmask_type>(num_bitmasks, stream);
+      std::fill(bitmask.begin(), bitmask.end(), ~bitmask_type{0});
+      return bitmask;
+    }
+  }();
+
+  auto validity_it = cudf::detail::make_counting_transform_iterator(
+    0, [bitmask = host_bitmask.data()](auto bit_index) { return bit_is_set(bitmask, bit_index); });
+  // Return only filtered row groups based on predicate
+  auto const is_row_required = cudf::detail::make_host_vector_sync(
+    device_span<uint8_t const>(input_rows.data<uint8_t>(), input_rows.size()), stream);
+
+  // Return if all are required, or all are nulls.
+  if (input_rows.null_count() == input_rows.size() or
+      std::all_of(
+        is_row_required.cbegin(), is_row_required.cend(), [](auto i) { return bool(i); })) {
+    return {};
```

Comment:
This shouldn't be the output and we should instead return `byte_range_info` for all pages in all row groups.